tools/dtrace/nexus/trace-transactions.d

```diff
@@ -0,0 +1,59 @@
+#!/usr/sbin/dtrace -qs
+
+/* Trace all transactions to the control plane database with their latency */
+
+dtrace:::BEGIN
+{
+    printf("Tracing all database transactions for nexus PID %d, use Ctrl-C to exit\n", $target);
+}
+
+/*
+ * Record the start and number of statements for each transaction.
+ *
+ * We're only recording the initial transactions, which are not nested.
+ */
+diesel_db$target:::transaction-start
+/arg1 == 0/
+{
+    this->key = copyinstr(arg0);
+    ts[this->key] = timestamp;
+    n_statements[this->key] = 0;
+}
+
+/*
+ * When a query runs in the context of a transaction (on the same connection),
+ * bump the statement counter.
+ */
+diesel_db$target:::query-start
+/ts[copyinstr(arg1)]/
+{
+    n_statements[copyinstr(arg1)] += 1
+}
+
+/*
+ * As transactions complete, print the number of statements we ran and the
+ * duration.
+ *
+ * NOTE: The depth, arg1, is intentionally 1 here. When we first enter a
+ * non-nested transaction, the depth is 0, because there is no outer
+ * transaction. However, when the `transaction-done` probe fires, we are still
+ * in a transaction, in which case the depth is currently 1. It's decremented
+ * _after_ the transaction itself is committed or rolled back.
+ */
+diesel_db$targe:::transaction-done
+/arg1 == 1 && ts[copyinstr(arg0)]/
+{
+    this->key = copyinstr(arg0);
```

Comment:
Is there any information (Line of code, name of file, name of transaction, etc) that can help us identify which transaction is being kicked off here?

This script is vastly better than not having a script, but I think the labeling of these transactions will be critical to making sense of "where our cost is going"!

Reply:
There isn't here, no. This is a first-stop script that (1) shows how to use the probes, and (2) gives a 50k-foot view. It'd definitely be great to get more details, and I think you could start from this script to do that. One thing you _can_ definitely do without much modification is print the queries themselves that run in a transaction.

Unfortunately, the actual details you're asking about here, like stacks or line information, are going to be hard to get. The main reason is Tokio. In general, we're running everything through the `transaction_async()` method -- IIRC, that wraps the transaction closure in something like an `async move {}` block, and then schedules it on the blocking task pool using `tokio::task::spawn_blocking`. (I probably have the exact details wrong, but I believe the main point I'm making still applies.)

The problem here is that the stack of _Tokio task_ has almost no relationship to our actual Omicron code that spawned it. For example, here are all the stack traces we get when the `transaction-start` probe fires over 10 seconds:


```
root@oxz_nexus_65a11c18:~# dtrace -n 'diesel_db*:::transaction-start { @[ustack()] = count(); }' -n 'tick-10s { printa(@); exit(0); }' -p $(pgrep nexus)
dtrace: description 'diesel_db*:::transaction-start ' matched 1 probe
dtrace: description 'tick-10s ' matched 1 probe
CPU     ID                    FUNCTION:NAME
120  11223                        :tick-10s

              nexus`_ZN176_$LT$diesel_dtrace..DTraceTransactionManager$LT$C$GT$$u20$as$u20$diesel..connection..transaction_manager..TransactionManager$LT$diesel_dtrace..DTraceConnection$LT$C$GT$$GT$$GT$17begin_transaction17hc62c02e2a5673f57E+0x315
              nexus`_ZN102_$LT$tokio..runtime..blocking..task..BlockingTask$LT$T$GT$$u20$as$u20$core..future..future..Future$GT$4poll17h2bc11f605c64db68E+0xa2
              nexus`_ZN5tokio7runtime4task7harness20Harness$LT$T$C$S$GT$4poll17h71ee64523863d4fbE+0x98
              nexus`_ZN5tokio7runtime8blocking4pool5Inner3run17h68830718001fcd88E+0xe4
              nexus`_ZN3std3sys9backtrace28__rust_begin_short_backtrace17hac03f594d2f3b4f1E+0x3f
              nexus`_ZN4core3ops8function6FnOnce40call_once$u7b$$u7b$vtable.shim$u7d$$u7d$17heb34e241cac90362E+0x7e
              nexus`_ZN3std3sys3pal4unix6thread6Thread3new12thread_start17hcaa249d1c5c30adfE+0x1b
              libc.so.1`_thrp_setup+0x77
              libc.so.1`_lwp_start
               66

              nexus`_ZN176_$LT$diesel_dtrace..DTraceTransactionManager$LT$C$GT$$u20$as$u20$diesel..connection..transaction_manager..TransactionManager$LT$diesel_dtrace..DTraceConnection$LT$C$GT$$GT$$GT$17begin_transaction17hc62c02e2a5673f57E+0x315
              nexus`_ZN102_$LT$tokio..runtime..blocking..task..BlockingTask$LT$T$GT$$u20$as$u20$core..future..future..Future$GT$4poll17h2bc11f605c64db68E+0xa2
              nexus`_ZN5tokio7runtime4task7harness20Harness$LT$T$C$S$GT$4poll17h4c72f11648d24933E+0x98
              nexus`_ZN5tokio7runtime8blocking4pool5Inner3run17h68830718001fcd88E+0xe4
              nexus`_ZN3std3sys9backtrace28__rust_begin_short_backtrace17hac03f594d2f3b4f1E+0x3f
              nexus`_ZN4core3ops8function6FnOnce40call_once$u7b$$u7b$vtable.shim$u7d$$u7d$17heb34e241cac90362E+0x7e
              nexus`_ZN3std3sys3pal4unix6thread6Thread3new12thread_start17hcaa249d1c5c30adfE+0x1b
              libc.so.1`_thrp_setup+0x77
              libc.so.1`_lwp_start
               66

```

They are not at all illuminating, since they basically show Tokio's stack when it schedules the async task that our transaction is actually running inside of. This is a pretty fundamental limitation of the interaction between DTrace and Tokio. I've thought for a long time, but not very deeply, about how to get more insight -- ultimately I think we're either going to need to insert DTrace probes in Tokio (when it schedules or swaps out tasks, for example), or in the Rust language implementation of `async` / `await` itself.

It's also possible that we can make even more USDT probes in the `transaction_async` method or similar. That would at least get us the Rust call stack when we entered _that_ function, which we could then use to predicate these probes we're listing here.

That's a long way of saying, yeah I definitely agree with you that we could use more detail about what transactions are running here! I'm not sure how to do that yet, but I'd love to improve the situation :)

Reply:
For the sake of github comment history: I added https://github.com/oxidecomputer/omicron/pull/7248

With that, we can see the name of transactions using something like:

```
nexus_db_queries$target:::transaction-start                                                                 
{                                                                                                           
    this->key = copyinstr(arg0);                                                                            
    this->conn_id = json(this->key, "ok");                                                                  
    printf("starting transaction %s on %s\n", copyinstr(arg1), this->conn_id);                              
}
```

Reply:
I'll modify this to use the new start / done probes from #7248 when that merges to `main`.